src/realm/sort_descriptor.cpp

```diff
@@ -391,6 +403,37 @@ void FilterDescriptor::execute(const Table& table, KeyValues& key_values, const
     key_values = std::move(filtered);
 }
 
+template <typename T, typename U>
+inline int compare(const T& i, const T& j, const U& col)
+{
+    const ColKey col_key{col.col_key};
+    const auto table = col.table;
+    const auto has_interner = col_key.get_type() == col_type_Mixed || col_key.get_type() == col_type_String;
+
+    if (has_interner) {
+        const auto id_i = i.cached_string_id;
+        const auto id_j = j.cached_string_id;
+        if (id_i || id_j) {
+            const auto interner = table->get_string_interner(col_key);
+
+            if (id_i && id_j) {
+                return interner->compare(*id_i, *id_j);
+            }
+            else if (id_i) {
+                // we need to inverse the returned value, since we are comparing I vs J.
+                const auto ret = interner->compare(j.get_value().get_string(), *id_i);
```

Comment:
we can't assume that j is a string here just because i has a intern string id. For example, it will assert when sorting a mixed column that has a string vs an int. If there isn't a test for this, could you please add one?

Reply:
yes good catch, I changed the code but forgot to push. We do have a test already.

Reply:
Note: there are 2 cases:
1. we have got a string not yet compressed (that does not have an ID) and we compare against a compressed string (either mixed or normal string) 

2 we are comparing int vs string or string vs int (or any other type), in this case we need to extract the compressed string (we can't avoid this)

2 is more expensive than 1.